extide/gradle/src/org/netbeans/modules/gradle/NbGradleProjectImpl.java

```diff
@@ -174,14 +174,76 @@ private Lookup createBasicLookup(ProjectState state, GradleAuxiliaryConfigImpl a
                 state
         );
     }
+    
+    GradleProject getProjectInternal() {
+        synchronized (this) {
+            return project;
+        }
+    }
 
     public GradleProject getGradleProject() {
-        if (project == null) {
-            project = loadProject();
+        GradleProject p = getProjectInternal();
+        if (p != null) {
+            return p;
+        }
+        p = loadProject();
+        synchronized (this) {
+            // avod unnecessary project replacements.
+            if (project != null) {
+                return project;
+            }
+            project = p;
         }
-        return project;
+        return p;
     }
-
+    
+    /**
+     * Records a reloaded project instance. Indicates whether the project was replaced.
+     * The method refuses to replace a project that has been changed since the reload procedure
+     * started. Will also not replace the project, if the original has a better quality
+     * than the reloaded one.
+     * <p>
+     * Pass {@code original = null} to just replace the project without checking.
+     * 
+     * @param original the original project data instance.
+     * @param reloaded the reloaded data.
+     * @param reloadIfMismatch will fire new reload if the project differs from original or quality is not satisfactory.
+     * @return {@code null}, if the project was accepted. {@code Boolean.TRUE}, if the project has
+     * been changed independently; {@code Boolean.FALSE}, if the reloaded project's quality is lower than
+     * the current.
+     */
+    Boolean recordOrReloadProject(GradleProject original, GradleProject reloaded, boolean reloadIfMismatch) {
+        Boolean result = null;
```

Comment:
Why is `result` tri-state? Can it be named `mismatchOfOldAndNewProject`, right?

Reply:
tri-state was explained below. If seems reasonable I would just leave it as it is; javadoc documents all three possible return values.

Reply:
OK, so degrading after reload is permitted. That's broken with this implementation - will revisit. Thanks.

Reply:
@lkishalmi I've checked the code again, recalling what I was thinking about that month ago: the code was written in such a way, as there can be multiple `loadProject` calls (potentialy) that would lead to `impl.project` replacement. The logic in the `recordOrReloadProject` basically ensures that if a project has been replaced **since the loadProject started** (the caller needs to save a reference), the `impl.project` will not be replaced for a worse version. But this is just for **competing** reloads.

Imagine a user clicking on Resolve problems, which will not make the project just `FULL` (whatever reason), but at the same time the background reload task will load the project and reaches `FULL_ONLINE`. If the final calls to `recordOrReloadProject` are reordered (different parallel threads) in an unfortunate sequence, the FULL_ONLINE read `impl.project` first, then it would degrade to FULL unnecessarily. The method's logic prevents that: if two replace processes are competing (caller has recorded `project` instance and passed it), the 'better of them' wins. Similar to AtomicReference.compareAndSet().

To be conservative, I've enabled this logic just from `GradleProjectProblemProvider` call. The other invocations pass `null` for `original`, so the comparison is disabled. So Reload action, reload based on desired quality change will always replace the project: I did not want to affect those occurrences, since I don't fully understand their semantics. I thought that it would not be good if "priming build" actually degraded already well-loaded project.

We can chat over the IM as you suggested or continue writing here. The logic to solve competing operations may be unnecessary.